File: leetcode/91_DecodeWays.cpp
//https://leetcode.com/problems/decode-ways/
#include<iostream>
#include<vector>
#include<algorithm>
#include<stdexcept>
using namespace std;
class Solution {
    public:
        int numDecodings(string s) {
            if(s.empty())   return 0;
            if(s.length()==1 && s[0]!='0')
                return 1;
            int n = s.size(), i, temp;
            vector<int> answer(n+2, 0);
            answer[n] = 1;
            answer[n+1] = 0;
            for(i=n-1;i>=0;i--){
                switch(s[i]){
                    case '0':
                        answer[i] = 0;
                        break;
                    case '1':
                        answer[i] = answer[i+1] + answer[i+2];
                        break;
                    case '2':
                        answer[i] = answer[i+1] + ((s[i+1] <= '6') ? answer[i+2] : 0);
                        break;
                    default:
                        answer[i] = answer[i+1];
                }
            }
            return answer[0];
        }
};
/* Test cases:
 *
 * 7        1
 * 17       2
 * 212      3        2,1,2   21,2    2,12
 * 108      1 
 *
 * */





int main(){
    class Solution s;
    int answer = s.numDecodings("212");
    //cout<<"Solution\n";
    cout<<answer<<endl;
    answer = s.numDecodings("17");
    answer = s.numDecodings("108");
    cout<<answer<<endl;

    return 0;
}

File: leetcode/91_DecodeWays.py
#https://leetcode.com/problems/decode-ways/
'''
    Inp: 12
    Out: 2 (AB or L)

    Inp: 45
    Out: 1 (DE)
'''
class Solution(object):
    def numDecodings(self, s):
        if not s:   return 0
        i = 0
        n = len(s)
        numWays = 1
        while i < n:
            if s[i] == '1' and i+1 < n:
                numWays += 1 if s[i+1] in '123456789' else 0
            elif s[i] == '2':
                if i+1 < n:
                    numWays += 1 if s[i+1] in '123456' else 0
                    i+=1
                i+=1
            elif s[i] == '0':
                return 0
            i+=1
        return numWays

s = Solution()
print s.numDecodings('12')
print s.numDecodings('45')
